book/go/api/writing-your-own-application.md

```diff
@@ -0,0 +1,184 @@
+
+For this, two things are needed:
+
+1. An entry point for Wallaroo to create the application. This is the function `ApplicationSetup()` that you need to define.
+2. The actual topology `ApplicationSetup()` is going to return for Wallaroo to create the application.
+
+#### Application Builder and Pipelines
+
+An application is constructed of pipelines which, in turn, are constructed from a sequence of a source, steps, and optionally a sink. Our reverse application only has one pipeline, so we only need to create one:
+
+```go
+application := app.MakeApplication("Reverse Word")
+application.NewPipeline("Reverse", app.MakeTCPSourceConfig(inHost, inPort, &Decoder{})).
+```
+
+Each pipeline must have a source, and each source must have a decoder, so `NewPipeline` takes a name and a `TCPSourceConfig` instance as its arguments.
+
+Next, we add the computation step:
+
+```go
+To(&ReverseBuilder{}).
+```
+
+And finally, we add the sink, using a `TCPSinkConfig`:
+
+```go
+ToSink(app.MakeTCPSinkConfig(outHost, outPort, &Encoder{}))
+```
+
+### The `ApplicationSetup` Entry Point
+
+After Wallaroo starts, it will try to execute its `ApplicationSetup()` function. This function is where the application builder steps from above are going to run.
+
+```go
+//export ApplicationSetup
+func ApplicationSetup() *C.char {
+  fs := flag.NewFlagSet("wallaroo", flag.ExitOnError)
+  inHostsPortsArg := fs.String("in", "", "input host:port list")
+  outHostsPortsArg := fs.String("out", "", "output host:port list")
+
+  fs.Parse(wa.Args[1:])
+
+  inHostsPorts := hostsPortsToList(*inHostsPortsArg)
+
+  inHost := inHostsPorts[0][0]
+  inPort := inHostsPorts[0][1]
+
+  outHostsPorts := hostsPortsToList(*outHostsPortsArg)
+  outHost := outHostsPorts[0][0]
+  outPort := outHostsPorts[0][1]
+
+  wa.Serialize = Serialize
+  wa.Deserialize = Deserialize
+
+  application := app.MakeApplication("Reverse Word")
+  application.NewPipeline("Reverse", app.MakeTCPSourceConfig(inHost, inPort, &Decoder{})).
+    To(&ReverseBuilder{}).
+    ToSink(app.MakeTCPSinkConfig(outHost, outPort, &Encoder{}))
+
+  return C.CString(application.ToJson())
+}
+```
+
+Configuration objects are used to pass information about sources and sinks to the application builder.
+
+Wallaroo provides convenience the functions `tcp_parse_input_addrs` and `tcp_parse_output_addrs` to parse host and port information that is passed on the command line, or the user can supply their own code for getting these values. When using the convenience functions, host/port pairs are represented on the command line as colon-separated values and multiple host/port values are represented by a comma-separated list of host/port values. The functions assume that `--in` is used for input addresses, and `--out` is used for output addresses. For example, this set of command line arguments would specify two input host/port values and one output:
```

Comment:
We don't provide `tcp_part_input_addrs` and `tcp_parse_output_addrs` functions in Go. The user is responsible for parsing their own command line input.

Reply:
fixed